app/services/foreman/provision/ssh.rb

```diff
@@ -77,32 +77,15 @@ def initiate_connection!
         Timeout.timeout(8) do
           ssh.run('pwd')
         end
-      rescue Errno::ECONNREFUSED
-        logger.debug "Connection refused for #{address}, retrying"
-        sleep(2)
-        retry
-      rescue Errno::EHOSTUNREACH
-        logger.debug "Host unreachable for #{address}, retrying"
-        sleep(2)
-        retry
-      rescue Net::SSH::Disconnect
-        logger.debug "Host dropping connections for #{address}, retrying"
-        sleep(2)
-        retry
-      rescue Net::SSH::ConnectionTimeout
-        logger.debug "Host timed out for #{address}, retrying"
-        sleep(2)
-        retry
-      rescue Net::SSH::AuthenticationFailed
-        logger.debug "Auth failed for #{username} at #{address}, retrying"
+      rescue => e
+        logger.info "An error occured while connecting before timeout occured \"#{e.inspect}\", retrying"
+        logger.debug "Full stacktrace of exception: \n  #{e.backtrace.join("\n  ")}"
         sleep(2)
         retry
-      rescue Timeout::Error
```

Comment:
Do we want to log also the `Timeout::Error`?

Reply:
By doing rescue e and then logging the exception class in the message we will know if it was a timeout or not

Reply:
I think retry is acceptable but we need to limit how many times we try retry. +1 to the rest.

Reply:
Number of retries is limited by the outer timeout block

Reply:
Never ever eat backtraces, I do not agree with @iNecas here and I would like to see it. If you think that could be too veerbose, just make sure it is logged only once (the first try). In the end, it's debug level (turned off by default).

Reply:
If the log contains trace of last exception, I'm ok with that. Especially if that's expected exception from which we recover by retry. But I don't insist :-)

Reply:
I think debug trace for known exceotions is ok: what I'm advocating for though is logging unknown exceptions AND backtrace with error log level

Reply:
If any exception is not rethrown, it must be logged. Previously we had particular exceptions and it was ok to swallow them, not now when we can miss a generic exception and not have stracktrace in logs.

Reply:
As a user, what I care about is that I can SSH into my box in this case I don't care what error occurs I just want it to retry for 5 minutes until it gets in. If it does time out its easy enough to enable debug logging and try it again. Theoretically (in this case for ssh) if the issue happens the first time and I turn on debug logs to get the stack trace I will get the same issue again. And in most cases the issue will not be on Foreman's side it will be something else with the server it is SSHing into.